jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourceInputStream.java

```diff
@@ -89,10 +96,21 @@ public int available() throws IOException
     @Override
     public void close()
     {
+        if (chunk == null)
+            chunk = content.read();
+
+        if (chunk != null && chunk.isLast())
+        {
+            if (!chunk.isTerminal())
+                chunk.release();
+            return;
+        }
+
         if (chunk != null)
             chunk.release();
-        if (chunk == Content.Chunk.EOF || chunk instanceof Content.Chunk.Error)
-            return;
-        chunk = Content.Chunk.EOF;
+
+        Throwable closed = new IOException("closed before EOF");
+        chunk = Content.Chunk.from(closed);
+        content.fail(closed);
```

Comment:
```suggestion
        if (chunk == Content.Chunk.EOF || chunk instanceof Content.Chunk.Error)
            return;
        Throwable closed = new IOException("closed before EOF");
        chunk = Content.Chunk.from(closed);
        content.fail(closed);
```
I think a combination of old code and new code is better.

Reply:
@sbordet the problem with your version is that it tests for `== EOF`, thus it will not see a normal chunk with isLast()==true as end of file and will terminate with a failure.

So my version is trying to say, that if a last chunk has been read, then a close is perfectly OK even if that chunk has unconsumed content... we just have to release it.   
I'll try to make that more explicit....

Reply:
Ah I didn't see that isTerminal checks for no remaining bytes, so my code is kind of wrong?   Or maybe not, should closing a non terminal last chunk be a failure?